In [1]:
import pandas as pd
import kagglehub
import matplotlib.pyplot as plt
import os

/mnt/hddtheo/EPFL/MA4/dviz/com-480-TnT/.venv/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


## Getting the dataset

In [2]:
# Download latest version of the Spotify Charts dataset
charts_path = kagglehub.dataset_download("dhruvildave/spotify-charts")

print("Path to dataset files:", charts_path)

Path to dataset files: /home/thehoul/.cache/kagglehub/datasets/dhruvildave/spotify-charts/versions/85


In [3]:
charts_csv_name = os.listdir(charts_path)[0]

In [4]:
charts_pd = pd.read_csv(charts_path + "/" + charts_csv_name)

In [5]:
charts_pd.head()

title  rank        date  \
0      Chantaje (feat. Maluma)     1  2017-01-01   
1  Vente Pa' Ca (feat. Maluma)     2  2017-01-01   
2   Reggaetón Lento (Bailemos)     3  2017-01-01   
3                       Safari     4  2017-01-01   
4                  Shaky Shaky     5  2017-01-01   

                                  artist  \
0                                Shakira   
1                           Ricky Martin   
2                                   CNCO   
3  J Balvin, Pharrell Williams, BIA, Sky   
4                           Daddy Yankee   

                                                 url     region   chart  \
0  https://open.spotify.com/track/6mICuAdrwEjh6Y6...  Argentina  top200   
1  https://open.spotify.com/track/7DM4BPaS7uofFul...  Argentina  top200   
2  https://open.spotify.com/track/3AEZUABDXNtecAO...  Argentina  top200   
3  https://open.spotify.com/track/6rQSrBHf7HlZjtc...  Argentina  top200   
4  https://open.spotify.com/track/58IL315gMSTD37D...  Argentina  top200   

           trend   streams  
0  SAME_POSITION  253019.0  
1        MOVE_UP  223988.0  
2      MOVE_DOWN  210943.0  
3  SAME_POSITION  173865.0  
4        MOVE_UP  153956.0

## Exploring the dataset

This dataset is supposed to contain the collection of each spotify chart published since january 1st 2017. 

In [6]:
# Print the number of entries for each date and region
print("Number of entries for each date and region:")
print(charts_pd.groupby(['date', 'region']).size())

Number of entries for each date and region:
date        region              
2017-01-01  Andorra                  25
            Argentina               250
            Australia               250
            Austria                 250
            Belgium                 250
                                   ... 
2021-12-31  United Arab Emirates    250
            United Kingdom           50
            United States           250
            Uruguay                  50
            Vietnam                  50
Length: 118072, dtype: int64


As we can see, the value is not always the same. Typically on january 1st 2017, Andorra has only 25 songs where as others have 250. Let's try to understand why.

In [7]:
andorra = charts_pd[charts_pd['region'] == 'Andorra']
andorra_2017_01_01 = andorra[andorra['date'] == '2017-01-01']
print("Andorra data for 2017-01-01:")
andorra_2017_01_01

Andorra data for 2017-01-01:


title  rank        date  \
21239753                         Reggaetón Lento (Bailemos)     1  2017-01-01   
21239754                                               Ride     2  2017-01-01   
21239755                   With Or Without You - Remastered     3  2017-01-01   
21239756                                      Feliz Navidad     4  2017-01-01   
21239757                                8 Days of Christmas     5  2017-01-01   
21239758                           Antes de que cuente diez     6  2017-01-01   
21239759                  19 Dias y 500 Noches - En Directo     7  2017-01-01   
21239760                                     Through - Live     8  2017-01-01   
21239761                                  Aqui No Hay Playa     9  2017-01-01   
21239762                                           Princesa    10  2017-01-01   
21239763                                    Amores de barra    11  2017-01-01   
21239764                      Ignition / Do You... (Mashup)    12  2017-01-01   
21239765                                            Píntame    13  2017-01-01   
21239766  Hopelessly Devoted To You - From “Grease” Soun...    14  2017-01-01   
21239767                                      Ay Que Pesado    15  2017-01-01   
21239768                                    Melt - Acoustic    16  2017-01-01   
21239769                                  All I Want Is You    17  2017-01-01   
21239770                                       Like Eye Did    18  2017-01-01   
21239771                                 I Hope You Find It    19  2017-01-01   
21239772                                             Correr    20  2017-01-01   
21239773                                           Brujeria    21  2017-01-01   
21239774                                 "Pa" Los Flamencos    22  2017-01-01   
21239775                                        A mi manera    23  2017-01-01   
21239776                                      Woman's World    24  2017-01-01   
21239777                   A Mi Manera - Tema Remasterizado    25  2017-01-01   

                                artist  \
21239753                          CNCO   
21239754             Twenty One Pilots   
21239755                            U2   
21239756                José Feliciano   
21239757               Destiny's Child   
21239758             Fito y Fitipaldis   
21239759                Joaquín Sabina   
21239760                George Michael   
21239761                 The Refrescos   
21239762    Joaquin Sabina y Viceversa   
21239763               Ella Baila Sola   
21239764                   Phoebe Ryan   
21239765                  Elvis Crespo   
21239766            Olivia Newton-John   
21239767                        Mecano   
21239768                         JONES   
21239769                            U2   
21239770                   FIL BO RIVA   
21239771                          Cher   
21239772                         Lewin   
21239773  El Gran Combo De Puerto Rico   
21239774                     El Barrio   
21239775                       Raphael   
21239776                          Cher   
21239777             Vicente Fernández   

                                                        url   region    chart  \
21239753  https://open.spotify.com/track/3AEZUABDXNtecAO...  Andorra  viral50   
21239754  https://open.spotify.com/track/2Z8WuEywRWYTKe1...  Andorra  viral50   
21239755  https://open.spotify.com/track/4kAJawkLZj4Q7N5...  Andorra  viral50   
21239756  https://open.spotify.com/track/5rgStA2VSw8K6jr...  Andorra  viral50   
21239757  https://open.spotify.com/track/2CZWUmcG9swKChE...  Andorra  viral50   
21239758  https://open.spotify.com/track/3xiNRrrVROKlHrf...  Andorra  viral50   
21239759  https://open.spotify.com/track/3LXFi1Xk9qJsYXY...  Andorra  viral50   
21239760  https://open.spotify.com/track/6B8vO1gto58lY9c...  Andorra  viral50   
21239761  https://open.spotify.com/track/6P1Lksu4JnHZhxw...  Andorra  viral50   
21239762  https://open.spotify.com/track/0yc

For some reason, only the top 25 songs are available for Andorra at this date. Let's look at other dates in that year.

In [8]:
# Let's create a year column from the date column
charts_pd['year'] = pd.to_datetime(charts_pd['date']).dt.year
# Get songs in Andorra in 2017
andorra_2017 = charts_pd[(charts_pd['region'] == 'Andorra') & (charts_pd['year'] == 2017)]
# Print the number of entries for each date
print("Number of entries for each date in Andorra in 2017:")
print(andorra_2017.groupby('date').size())
# Print the different counts of entries for each date
print("Different counts of entries for each date in Andorra in 2017:")
print(andorra_2017.groupby('date').size().value_counts())

Number of entries for each date in Andorra in 2017:
date
2017-01-01    25
2017-01-02    25
2017-01-03    25
2017-01-04    22
2017-01-05    16
              ..
2017-12-15    50
2017-12-16    50
2017-12-17    50
2017-12-18    50
2017-12-19    50
Length: 276, dtype: int64
Different counts of entries for each date in Andorra in 2017:
50    87
16    11
2      9
1      9
32     8
15     7
34     6
25     6
17     6
10     6
30     6
29     6
42     6
18     6
35     5
20     5
27     5
26     5
13     4
41     4
43     4
19     4
31     4
33     4
14     4
8      4
23     4
28     3
4      3
40     3
22     2
38     2
12     2
45     2
9      2
46     2
49     2
24     2
11     2
5      2
37     2
3      2
47     2
36     1
6      1
7      1
39     1
21     1
48     1
Name: count, dtype: int64


Really strange. Some times we have 50 songs (30%) and the rest of the times only less. May be that the data for that country is not as complete as others. Let's check the United States which I expect to be the most complete.

In [9]:
us_2017 = charts_pd[(charts_pd['region'] == 'United States') & (charts_pd['year'] == 2017)]
# Print the number of entries for each date
print("Number of entries for each date in US in 2017:")
print(us_2017.groupby('date').size())
# Print the different counts of entries for each date
print("Different counts of entries for each date in US in 2017:")
print(us_2017.groupby('date').size().value_counts())

Number of entries for each date in US in 2017:
date
2017-01-01    250
2017-01-02    250
2017-01-03    250
2017-01-04    250
2017-01-05    250
             ... 
2017-12-27    250
2017-12-28    250
2017-12-29    250
2017-12-30    250
2017-12-31    250
Length: 365, dtype: int64
Different counts of entries for each date in US in 2017:
250    351
249      6
50       3
200      1
246      1
245      1
248      1
199      1
Name: count, dtype: int64


This is reassuring. The US has 250 songs for almost all dates (in 2017). Let's now check what these 250 songs are. I expect to have the 200 songs of the top200 and the 50 songs of the viral 50 (and thus have the 50 top songs of the top200 to be duplicates of the viral 50). Let's check that.

In [10]:
us_2017_01_01 = us_2017[us_2017['date'] == '2017-01-01']
us_2017_01_01[us_2017_01_01['rank'] == 1]

title  rank        date  \
8246      Bad and Boujee (feat. Lil Uzi Vert)     1  2017-01-01   
21254585         Fill Me In - AFTR:HRS Mixcut     1  2017-01-01   

                              artist  \
8246                           Migos   
21254585  Loud Luxury, Ryan Shepherd   

                                                        url         region  \
8246      https://open.spotify.com/track/4Km5HrUvYTaSUfi...  United States   
21254585  https://open.spotify.com/track/1Th5b8AsauED9F7...  United States   

            chart          trend    streams  year  
8246       top200  SAME_POSITION  1371493.0  2017  
21254585  viral50  SAME_POSITION        NaN  2017

After some research, I found out that the top200 and the viral50 are not the same. The top200 is a collection of the most streamed songs while the viral50 is a collection of the most "viral" songs, whatever that means.

This is nice! The rest of the dataset is straight forward. Every time a new chart is published, the "trend" column indicates for each song if it is a new entry, and if not, if it moved up or down in the chart. For each entry, the "streams" are indicated at that time. Other than this, we have the song name, artist and url. 

## Preparing the dataset

We are going to use this dataset to get information about how trends evolved over time. We will rely on the second dataset (Spotify tracks) to get audio features of the songs (danceability, energy, etc.). This second dataset will thus contain static metadata about the songs. This means that this dataset only need to contain the song's spotify id and we will query the audio features using this id in the tracks dataset.

Let's start by extracting the spotify id from the url and drop the url, title and artist columns.

In [11]:
charts_pd['id'] = charts_pd['url'].str.split('/').str[-1]
charts_pd.drop(columns=['url', 'title', 'artist'], inplace=True)

Let's now save this dataset as a pkl to be used later.

In [13]:
# Save the cleaned dataset to pkl
charts_pd.to_pickle("./spotify_charts_cleaned.pkl")